core-bundle/src/Migration/CommandCompiler.php

```diff
@@ -0,0 +1,200 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Migration;
+
+use Contao\CoreBundle\Doctrine\Schema\SchemaProvider;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Schema\Schema;
+
+class CommandCompiler
+{
+    /**
+     * @var array<string, string>|null
+     */
+    private array|null $commands = null;
+
+    /**
+     * @internal Do not inherit from this class; decorate the "contao.migration.command_compiler" service instead
+     */
+    public function __construct(private readonly Connection $connection, private readonly SchemaProvider $schemaProvider)
+    {
+    }
+
+    /**
+     * Returns a list of SQL commands indexed by a unique hash.
+     *
+     * @return array<string, string>
+     */
+    public function getCommands(): array
+    {
+        return $this->commands ?? ($this->commands = $this->compileCommands());
+    }
+
+    /**
+     * Execute a SQL command identified by its hash.
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function execCommand(string $hash): void
+    {
+        $command = $this->commands[$hash] ?? throw new \InvalidArgumentException(sprintf('Invalid hash: %s', $hash));
+
+        $this->connection->executeStatement($command);
+    }
```

Comment:
We could IMHO get rid of this function and also get rid of the caching of commands. IMHO this would be an even cleaner API. Wdyt?

Reply:
Yes :+1: for removing this, removing the caching and removing the md5 hashes.

Reply:
see 4d2f0ee64c56a44b63deb6976be012af8913bc22

Reply:
Now that the md5 hashes are an implementation detail, we could likely also refactor the migrate command. But maybe in another PR?

Reply:
I’d do it in this PR, should be straight forward I think?
A quick and dirty way would be to just replace `$commands[md5($command)] = $command;` with `$commands[$command] = $command;` 😅

Reply:
Ok, I'll look into it. This will change how the `sha1` will look, but they weren't designed to be persisted, correct?

Reply:
and probably so that the `Input` class does’t mangle the values ☺️